### PR TITLE
[#45] gitignore

### LINKED ISSUE #45: add .gitignore file for all project configurations
add things like:

```
node_modules
dist
.cache
```

### PR DESCRIPTION
**Pull request by jacekk**
Fixes #45 . Ignore a lot, but IMHO highly improbable it is too much.

### CODE REVIEW

**Comment by jakoblind** [comment]
I'm thinking about what is best practice. When I start a new project I like to create a minimal `.gitignore` and only update with files that I want to ignore. In this `.gitignore` that you propose, it's another strategy, adding lots of files even if they are currently not needed.

For example, ignoring the following is not needed right away. (but might be in the future)
```
*.com
*.class
*.dll
*.exe
*.o
*.so
```

What I'm worried about is that for new people it can be overwhelming to see all these lines.

**Comment by jacekk** [comment]
@jakoblind I prefer more cautious attitude - seen many OS-generated files/directories under GIT - but I get your point. Updated. Looks better now?

File: src/templates/base.js [hunk 1]
```diff
@@ -123,3 +119,20 @@ ${
 Made with [createapp.dev](https://createapp.dev/)
 
 `;
+
+export const gitignore = () => `
+.cache/
+coverage/
+dist/
```
Comment (jakoblind) [inline src/templates/base.js]: currently, the `index.html` file is placed inside `dist` folder. Maybe we should consider changing that. But in the meantime, let's not add `dist/` to `.gitignore`

Comment (jacekk) [reply src/templates/base.js]: @jakoblind 
I can see 2 "more proper" options:

1. add `!dist/index.html` right after `dist/` into the `.gitignore` file;
1. use [this webpack plugin](https://github.com/jantimon/html-webpack-plugin) to omit keeping any source/versioned files within `dist` directory; it should be safely removable

Comment (jakoblind) [reply src/templates/base.js]: I was thinking to add `html-webpack-plugin` as a selectable feature. Until then let's go for your option 1. That's nice and explicit.

Comment (jacekk) [reply src/templates/base.js]: @jakoblind Added.

**Comment by jakoblind** [comment]
image

I think the `.gitignore` should be on top of the file list. Also, the syntax highlighting looks weird. Let me know if you need any guidance on this

**Comment by jacekk** [comment]
@jakoblind I've never used PrismJS and this project setup seems kinda weird. I was able to more syntaxes in this PR - https://github.com/jacekk/webpack-autoconf/pull/1/files - but I am not conviced this is the "proper" way of using that library in the context of this very project. Yup, guidance is welcome :)

**Comment by jakoblind** [comment]
I think it should work to only change the following line in `FileBrowser.js:102`

` <code className="language-javascript">{code}</code>`

To `language-css`, etc. Doesn't that work for you? 

I liked your solution to properly apply the corrrect syntax highlighting based on extension!

**Comment by jacekk** [comment]
@jakoblind 
Well, that lines would work, but only for CSS files. What about other files types? There are json, tsx, md, vue, etc., generated by this tool. Do we intentionally skip support for those? When I've console.log'ged `Prism.languages` before any changes, the list was very short. Certainly, there was no JSON, not MD among them. You can run `Object.keys(Prism.languages)` on https://createapp.dev/ to find out.

P.S.
Prism does not support `vue` nor `svelte` syntaxes at all. I've set the language as `markup` - talking about my branch changes - cause it looks good enough then.

**Comment by jakoblind** [comment]
I see your point. 

I read on the [prism docs](https://prismjs.com/#basic-usage) that they recommend using the [babel-plugin-prismjs](https://github.com/mAAdhaTTah/babel-plugin-prismjs) when using npm to define what languages should be supported. we should try that.

Although we should use the highlight plugin with `babel-plugin-prismjs` because we're using a custom version of it right now.

**Comment by jacekk** [comment]
@jakoblind 
I definetely want to play around with that babel plugin. My question, is this related to this PR / issue scope? I believe not, so I would vote to not care about `.gitignore` highlighting, as it could be improved/fixed in a separate PR. Do you agree?

**Comment by jakoblind** [comment]
I agree. What we can do as a short term fix is to disable highlighting for the `.gitignore` file. It's not much to highlight anyway

**Comment by jacekk** [comment]
@jakoblind 
Updated to highlight those file types which are supported in current/attached version of Prism. `.gitignore` looks better, but also other files looks better in a non-hardcoded-javascript mode. Looks good enough?

File: src/components/FileBrowser.js [hunk 2]
```diff
@@ -162,7 +178,8 @@ class FileBrowser extends React.Component {
     const { fileContentMap } = this.props;
     const fileContent = _.get(fileContentMap, this.state.selectedFile, '');
 
-    const extension = this.state.selectedFile.match(extensionRegex);
+    const extensionMatch = this.state.selectedFile.match(extensionRegex);
+    const extension = Array.isArray(extensionMatch) ? extensionMatch[0] : null;
```
Comment (jakoblind) [inline src/components/FileBrowser.js]: We use lodash in this project, consider write this in lodash for consitency. By the way, is there a eslint rule for this?

Comment (jacekk) [reply src/components/FileBrowser.js]: @jakoblind 
Used lodash version if this function in 1f8570f651daa666bc31f436cb462f8c0e8142d7 . BTW, `extensionMatch ? extensionMatch[0] : null;` should work exactly the same in this case, dunno why I like `isArray` ;)

Comment (jacekk) [reply src/components/FileBrowser.js]: @jakoblind 
Re eslint; just found sth like this: https://github.com/wix/eslint-plugin-lodash/blob/HEAD/docs/rules/prefer-lodash-method.md , but I dunno whether you were thinking about sth like that.

Comment (jakoblind) [reply src/components/FileBrowser.js]: Another option is to use [_.head](https://lodash.com/docs/4.17.15#head). It returns `undefined` if it's not an array, or an empty array. I think it's a bit more readable, but you decide.

Yes those eslint rules was what I was thinking about! Will look into that

Comment (jacekk) [reply src/components/FileBrowser.js]: @jakoblind 
Like it. Updated.

**Pull request merged**